### PR TITLE
Test262 integration

### PR DESCRIPTION
**Pull request by Protryon**

### CODE REVIEW

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java
Comment (bakkot) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: I am morally opposed to skipped tests. They should be xfailed, not skipped. Otherwise the expectations file gets stale as new features are implemented.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java
Comment (bakkot) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: I think we probably just want to depend on a yaml parser.

Comment (Protryon) [reply src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: Added snakeyaml

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java
Comment (Protryon) [reply src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: Removed all skipped tests in favor of xfailing

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java
Comment (michaelficarra) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: This isn't true, these are bugs with our handling of `yield` in various contexts.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java
Comment (michaelficarra) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: Can you dedupe the test path? These should all start with `src/test/resources/test262/test` so extract that. This will make it easier in case we want to change the location of our test262 submodule or something.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java
Comment (michaelficarra) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: Again, these should just be considered bugs.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java
Comment (michaelficarra) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: Not in ES2016.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java
Comment (michaelficarra) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: More importantly, not in ES2016, the version supported in this branch.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java
Comment (michaelficarra) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: Can't we filter these based on feature flag? We should whitelist/blacklist supported/unsupported feature flags.

Comment (Protryon) [reply src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: No `async` feature flag exists, the flag is much too broad. Trailing commas, etc, do not have flags or feature flags to accompany them.

Comment (bakkot) [reply src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: @Protryon There are `async-functions` and `async-iteration` feature flags which cover most of them (and the exceptions are arguably bugs in the test suite). That's [what shift-parser-js does](https://github.com/shapesecurity/shift-parser-js/blob/f973b067adbb7e72fe3a010596bc50c73a5e6f3c/test/test262/expectations/index.js#L13-L14).

I'm surprised this looks so different from the shift-parser-js tests, in fact. For example, there's [well over a hundred trailing comma tests](https://github.com/shapesecurity/shift-parser-js/blob/f973b067adbb7e72fe3a010596bc50c73a5e6f3c/test/test262/expectations/trailing-function-commas.js#L174), but I only see reference to 14 here.

Comment (Protryon) [reply src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: @bakkot `async-functions` covers 4 tests, and `async-iteration` covers none (doesn't exist), at least in the commit currently in the repo. As for all the trailing comma tests, they are running and passing. I suppose that in itself could be a bug in the parser, however IIRC trailing commas are in es2016, just not for function arguments or calls. There is no feature flag for trailing commas, however a hack would be to check if the name contains `trailing-comma`.

Comment (bakkot) [reply src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: @Protryon ah, that's because the test262 you're using is two years out of date. It should probably be updated - try ` 72f1cfa`.

(Yes, that will mean pulling in tests for new features, but that's OK. And it matches what shift-parser-js does.)

Comment (bakkot) [reply src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: > There is no feature flag for trailing commas, however a hack would be to check if the name contains trailing-comma.

Just list them out manually, [like shift-parser-js](https://github.com/shapesecurity/shift-parser-js/blob/f973b067adbb7e72fe3a010596bc50c73a5e6f3c/test/test262/expectations/trailing-function-commas.js). I would expect this project's test expectations to look almost exactly like shift-parser-js's.

**Comment by bakkot** [comment]
@Protryon Let's roll test262 before landing this.

File: src/main/java/com/shapesecurity/shift/es2016/parser/GenericParser.java
Comment (bakkot) [inline src/main/java/com/shapesecurity/shift/es2016/parser/GenericParser.java]: This looks to me like it would reject `(-1) ** 2`. Do we have a test covering that case? If not we should add one.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java
Comment (bakkot) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: I don't like defaults here. We should throw if we encounter something we don't recognize.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java
Comment (bakkot) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: I would prefer more informative messages. For example, this message does not make it clear if the problem is that the test was xfailed and passed, or was not xfailed but failed.

I think each of the four failure cases (where `[ xfailed, shouldFail, failed ]` takes values `[ false, false, true ]`, `[ false, true, false ]`, `[ true, false, false ]`, `[ true, true, true ]`) warrants its own error message.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java
Comment (bakkot) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: I would prefer we explicitly skip fixtures, and then require the frontmatter to parse. That way we don't accidentally pass a test just because its frontmatter is in a format we don't recognize.

Per [INTERPRETING.md](https://github.com/tc39/test262/blob/dec4e5df7481daf086036a152abb30492dec398a/INTERPRETING.md#modules), fixtures are those which end with `_FIXTURE.js`. We can just ignore those files.

**Review by bakkot — APPROVED** [review]
LGTM modulo comments above.

File: src/main/java/com/shapesecurity/shift/es2016/parser/GenericParser.java
Comment (Protryon) [reply src/main/java/com/shapesecurity/shift/es2016/parser/GenericParser.java]: Test added, passing

**Comment by Protryon** [comment]
@bakkot Addressed comments

File: src/test/java/com/shapesecurity/shift/es2016/parser/expressions/BinaryExpressionTest.java
Comment (bakkot) [inline src/test/java/com/shapesecurity/shift/es2016/parser/expressions/BinaryExpressionTest.java:191]: Given the approach you've taken, can we also add a test `++a**2` does parse (as `(++a)**2`)?

Comment (Protryon) [reply src/test/java/com/shapesecurity/shift/es2016/parser/expressions/BinaryExpressionTest.java:191]: @bakkot Added

**Pull request merged**
